runtime/near-vm-runner/src/runner.rs

```diff
@@ -8,43 +10,53 @@ use near_vm_errors::{CompilationError, FunctionCallError, VMError};
 use near_vm_logic::types::PromiseResult;
 use near_vm_logic::{External, VMContext, VMKind, VMOutcome};
 
-/// `run` does the following:
-/// - deserializes and validate the `code` binary (see `prepare::prepare_contract`)
-/// - injects gas counting into
-/// - adds fee to VMLogic's GasCounter for size of contract
-/// - instantiates (links) `VMLogic` externs with the imports of the binary
-/// - calls the `method_name` with `context.input`
-///   - updates `ext` with new receipts, created during the execution
-///   - counts burnt and used gas
-///   - counts how accounts storage usage increased by the call
-///   - collects logs
-///   - sets the return data
-///  returns result as `VMOutcome`
-pub fn run<'a>(
-    code: &ContractCode,
-    method_name: &str,
-    ext: &mut dyn External,
-    context: VMContext,
-    wasm_config: &'a VMConfig,
-    fees_config: &'a RuntimeFeesConfig,
-    promise_results: &'a [PromiseResult],
-    current_protocol_version: ProtocolVersion,
-    cache: Option<&'a dyn CompiledContractCache>,
-    profile: &ProfileData,
-) -> (Option<VMOutcome>, Option<VMError>) {
-    run_vm(
-        code,
-        method_name,
-        ext,
-        context,
-        wasm_config,
-        fees_config,
-        promise_results,
-        VMKind::default(),
-        current_protocol_version,
-        cache,
-        profile.clone(),
-    )
+pub struct WasmMachine {
+    vm_config: VMConfig,
+    cache: Option<Arc<dyn CompiledContractCache>>,
+}
+
+impl WasmMachine {
+    pub fn new(vm_config: VMConfig, cache: Option<Arc<dyn CompiledContractCache>>) -> WasmMachine {
+        WasmMachine { vm_config, cache }
+    }
+
+    /// `run` does the following:
+    /// - deserializes and validate the `code` binary (see `prepare::prepare_contract`)
+    /// - injects gas counting into
+    /// - adds fee to VMLogic's GasCounter for size of contract
+    /// - instantiates (links) `VMLogic` externs with the imports of the binary
+    /// - calls the `method_name` with `context.input`
+    ///   - updates `ext` with new receipts, created during the execution
+    ///   - counts burnt and used gas
+    ///   - counts how accounts storage usage increased by the call
+    ///   - collects logs
+    ///   - sets the return data
+    ///  returns result as `VMOutcome`
+    pub fn run(
+        &mut self,
+        code: &ContractCode,
+        method_name: &str,
+        ext: &mut dyn External,
+        context: VMContext,
+        fees_config: &RuntimeFeesConfig,
+        promise_results: &[PromiseResult],
+        current_protocol_version: ProtocolVersion,
+        profile: &ProfileData,
+    ) -> (Option<VMOutcome>, Option<VMError>) {
+        run_vm(
```

Comment:
> The goal here is to re-use VM resources between contract executions. At
minimum, we would want to re-use virtual memory mapping crated by wasm
runtimes. 

This PR so far looks good. However `machine.run` still doesn't reuse wasmer instance and wasmer memory, which is the major thing common to any contract execution, that should consider storing them in machine to reuse. And I think we should do a quick benchmark to see how fast is `virtual memory mapping crated by wasm
runtimes` and init wasmer/wasmtime instance, if it's quicker than cleanup existing wasm memory modules, or if it's trivial as bottleneck, then `WasmerMachine {vm_config,cache}` in this PR is sufficient

Reply:
Yeah, this PR intentionally avoids implementing any actual re-use, to keep it smaller and to just establish the API.